opal/datatype/opal_copy_functions_heterogeneous.c

```diff
@@ -141,106 +592,183 @@ static inline void opal_dt_swap_long_double(void *to_p, const void *from_p, cons
  */
 #define COPY_TYPE_HETEROGENEOUS(TYPENAME, TYPE) COPY_TYPE_HETEROGENEOUS_INTERNAL(TYPENAME, TYPE, 0)
 
+#if 0
+Summaryizing the logic of the pFunc copy functions
+with regard to long doubles:
+
+For terminology I''ll use
+f64 : float64 which some architectures use as their long double
+f80 : x86 double extended format that uses 80 bytes, commonly used for long double
+f128 : ieee quad precision, sometimes available as __float128
+
+  if !LONG_DOUBLE or both architecture have the same long double format:
+    byte swap based on local/remote endianness differing
+  else:
+    if from_arch is not local endianness: byte swap to local endianness
+    if from_arch isn''t f128 : ldbl_to_f128
+      if we have __float128         : convert to __float128
+      else if from_arch LDBL is f80 : f80_to_f128
+      else if from_arch LDBL is f64 : f64_to_f128
+    if to_arch isn''t f128 : f128_to_ldbl
+      if we have __float128         : convert from __float128 to 
+      if to_arch LDBL is f80   : f128_to_f80
+      if to_arch LDBL is f64   : f128_to_f64
+    if to_arch is not local endianness : byte swap
+
+And for all the above conversions the logic for handling size difference
+between the from/to type is the same:
+  if (to_extent == from_extent == sizeof(TYPE))
+    opal_dt_swap_bytes(to, from, sizeof(TYPE), count);
+  else
+    loop i=0..count-1
+      opal_dt_swap_bytes(to, from, sizeof(TYPE), 1);                                 \
+      to += to_extent;
+      from += from_extent;
+so that''s handled by a do while as an outer loop.
+#endif
+
 #define COPY_TYPE_HETEROGENEOUS_INTERNAL(TYPENAME, TYPE, LONG_DOUBLE)                              \
     static int32_t copy_##TYPENAME##_heterogeneous(opal_convertor_t *pConvertor, size_t count,     \
```

Comment:
I don't understand why you need the `from_arch` and `to_arch` as they both should be defined in the convertor itself. Indeed the convertor knows if it is doing a pack or an unpack, as well as what is the source and destination architecture so why not use them instead of additional arguments ?

Reply:
the convertor converts between local architecture into the master->remote_arch, but the conversion itself only happens on one of the peers involved in the communication. The conversion is done during pack (if the convertor flag is set to CONVERTOR_SEND_CONVERSION) or during unpack (if the convertor flag is not set to CONVERTOR_SEND_CONVERSION), so knowing the type of the operation and the convertor flags you can deduce the from and to arch.

Reply:
Awesome, so I added this then in place of the new from_arch/to_arch arguments:
```
+        int from_arch, to_arch ;                                        \
+        if (pConvertor->flags & CONVERTOR_SEND_CONVERSION) { /* pack */ \
+            from_arch = opal_local_arch;                                \
+            to_arch = pConvertor->remoteArch;                           \
+        } else { /* unpack */                                           \
+            from_arch = pConvertor->remoteArch;                         \
+            to_arch = opal_local_arch;                                  \
+        }                                                               \
```
It seems to work, does it look okay?